lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -0,0 +1,394 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements FIRRTL module instance inlining.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Module Inlining Support
+//===----------------------------------------------------------------------===//
+
+/// If this operation or any child operation has a name, add the prefix to that
+/// operation's name.
+static void rename(const Twine &prefix, Operation *op) {
+  // TODO: this should really be an OpInterface.. Or we can just blindly
+  // modify the attribute "name" indiscriminantly.
+  llvm::TypeSwitch<Operation *>(op)
+      .Case<CMemOp, InstanceOp, MemOp, MemoryPortOp, NodeOp, RegOp, RegResetOp,
+            SMemOp, WireOp>([&](auto op) {
+        // TODO: one day names won't be optional
+        if (auto name = op.nameAttr()) {
+          std::string newName = (prefix + name.getValue()).str();
+          op.nameAttr(StringAttr::get(op.getContext(), newName));
+        }
+      });
+  // Recursively rename any child operations.
+  for (auto &region : op->getRegions())
+    for (auto &block : region)
+      for (auto &op : block)
+        rename(prefix, &op);
+}
+
+/// Clone an operation, mapping used values and results with the mapper, and
+/// apply the prefix to the name of the operation. This will clone to the
+/// insert point of the builder.
+static void cloneAndRename(const Twine &prefix, OpBuilder b,
+                           BlockAndValueMapping &mapper, Operation &op) {
+  auto newOp = b.clone(op, mapper);
+  rename(prefix, newOp);
+}
+
+/// This function is used before inlining a module, to handle the conversion
+/// between module ports and instance results. For every port in the target
+/// module, create a wire, and assign a mapping from each module port to the
+/// wire. When the body of the module is cloned, the value of the wire will be
+/// used instead of the module's ports.
+static SmallVector<Value> mapPortsToWires(const Twine &prefix, OpBuilder &b,
+                                          BlockAndValueMapping &mapper,
+                                          FModuleOp target) {
+  SmallVector<Value> wires;
+  auto portInfo = target.getPorts();
+  for (unsigned i = 0; i < target.getNumArguments(); ++i) {
```

Comment:
nit, no need to evaluate getNumArguments() each time through the loop

Reply:
Fixed x4